ntp-proto/src/packet/crypto.rs

```diff
@@ -0,0 +1,112 @@
+use aes_siv::{siv::Aes128Siv, Key, KeyInit};
+use rand::Rng;
+use tracing::error;
+
+use super::{extensionfields::ExtensionField, PacketParsingError};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EncryptionResult {
+    pub nonce: Vec<u8>,
+    pub ciphertext: Vec<u8>,
+}
+
+pub trait Cipher: Sync + Send {
+    fn encrypt(
+        &self,
+        plaintext: &[u8],
+        associated_data: &[u8],
+    ) -> std::io::Result<EncryptionResult>;
+    fn decrypt(
+        &self,
+        nonce: &[u8],
+        ciphertext: &[u8],
+        associated_data: &[u8],
+    ) -> Result<Vec<u8>, PacketParsingError>;
+}
+
+pub trait CipherProvider {
+    fn get(&self, context: &[ExtensionField<'_>]) -> Option<&dyn Cipher>;
+}
```

Comment:
none of the current impls use the `context` argument. Your assumption is that certain ciphers will in the future?

Reply:
This is for the server's CipherProvider, so that it can get access to the cookie it needs to get the key.